### ISSUE TITLE
SSHFactory legacy support of string keys is broken

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3141]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)| @therve reported|
|-|-|
|Trac ID|trac#3141|
|Type|release blocker: regression|
|Created|2008-03-31 15:03:28Z|
|Branch|https://github.com/twisted/twisted/tree/sshfactory-keys-3141|

It does make the check on self.publicKeys/self.privateKeys, but only if it comes from self.getPublicKeys/self.getPrivateKeys.

<details><summary>Searchable metadata</summary>

```
trac-id__3141 3141
type__release_blocker__regression release blocker: regression
reporter__therve therve
priority__normal normal
milestone__ 
branch__branches_sshfactory_keys_3141 branches/sshfactory-keys-3141
branch_author__ 
status__closed closed
resolution__fixed fixed
component__conch conch
keywords__ 
time__1206975808000000 1206975808000000
changetime__1207125606000000 1207125606000000
version__None None
owner__ 
cc__z3p
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

This is a change in behavior between 2.5.0 and 8.0.1.  Changing to the brand new `regression` type.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve set owner to @therve|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_3"></a>|@therve commented|
|-|-|

(In [23153]) Branching to 'sshfactory-keys-3141'

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_4"></a>|@therve commented|
|-|-|

Please review!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun set owner to @therve<br>@exarkun set status to `new`|
|-|-|

Thanks for fixing this.  Imaginary works again. :)  Some pretty minor things:

  * Is `unittest.__file__` the best value to pass for the filename parameter to `assertWarns`?  For normal two-frame deprecations (ie, the error is at the callsite of the deprecated function), I feel like it's better to define a function in the test method and use the test module's __file__ as the expected filename.  Maybe this isn't any better than using `unittest.__file__` though.  However, this doesn't seem like the normal two-frame case: the API issuing the deprecation will be invoked from `Factory.doStart`, which will be invoked from `tcp.Port.startListening`, which will be invoked from some reactor's `listenTCP` implementation, which _may_ be invoked from user code, or which may be invoked from `TCPServer.startService`, which will be invoked from some arbitrarily deeply nested `IService.startService` implementation, which will ultimately be invoked by some implementation detail in `twistd`.  So, ah, maybe stack frames don't make any sense in this case.  The FQPN of the `SSHFactory` subclass which is doing the deprecated thing would probably be more useful.
  * Maybe you can ignore this, but the `warned` variable in the loops that are changed here seems superfluous.  The Python warning system already eliminates duplicate warnings.  I wonder if I'm misunderstanding something about that code.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_7"></a>|@therve commented|
|-|-|

(In [23156]) Fix the assertWarns problem with multiple warnings, to be able to remove the
awkward code in SSHFactory.

Refs [#2904](https://github.com/twisted/twisted/issues/2904)
Refs [#3141](https://github.com/twisted/twisted/issues/3141)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_8"></a>|@therve removed owner|
|-|-|

Thanks for the review, I think that improves the code a lot. For the record, the warned variable was there because of a bug in assertWarns which wasn't able to handle the same warning multiple times (see [#2904](https://github.com/twisted/twisted/issues/2904)). I resolved that (I think).

Please review!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun set owner to @therve|
|-|-|

Looks good.  I'm beginning to wonder if `assertWarns` is really a very good API, but this is definitely an improvement.

Please merge.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_10"></a>|@therve set status to `closed`|
|-|-|

(In [23170]) Merge sshfactory-keys-3141

Author: therve
Reviewer: exarkun
Fixes [#3141](https://github.com/twisted/twisted/issues/3141)
Fixes [#2904](https://github.com/twisted/twisted/issues/2904)

Fix backward compatibility behavior of SSHFactory, allowing the publicKeys and
privateKeys attributes to be string instead of key objects. Fix also a problem
in assertWarns with multiple warnings.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
